### ISSUE TITLE
ImportError: cannot import name BPF

### ISSUE THREAD

**Issue by britchey80** [iovisor/bcc#2278]
4.9.0-8-amd64 #1 SMP Debian 4.9.144-3 (2019-02-02) x86_64 GNU/Linux

After following the installation instructions for Debian here:
https://github.com/iovisor/bcc/blob/master/INSTALL.md#debian---source

I am unable to run any of the tools.  They all fail with the error message:
```
from bcc import BPF
ImportError: cannot import name 'BPF'
```

Here are my current kernel BPF configs:
```
cat /boot/config-4.9.0-8-amd64 | grep BPF
CONFIG_BPF=y
CONFIG_BPF_SYSCALL=y
# CONFIG_BPF_JIT_ALWAYS_ON is not set
CONFIG_NETFILTER_XT_MATCH_BPF=m
CONFIG_NET_CLS_BPF=m
CONFIG_NET_ACT_BPF=m
CONFIG_BPF_JIT=y
CONFIG_HAVE_EBPF_JIT=y
CONFIG_BPF_EVENTS=y
CONFIG_TEST_BPF=m
```

Any help would be appreciated.

**Comment by BradWalker** [comment]
I also encounter this same issue on Ubuntu 18.04 when using the installed tools..

bwalker@brads-T460s:~$ apt list --installed | grep bpfcc-tools

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

bpfcc-tools/bionic,bionic,now 0.5.0-5ubuntu1 all [installed]
bwalker@brads-T460s:~$ 

bwalker@brads-T460s:~$ /usr/sbin/biolatency-bpfcc 
Traceback (most recent call last):
  File "/usr/sbin/biolatency-bpfcc", line 15, in <module>
    from bcc import BPF
ImportError: cannot import name BPF
bwalker@brads-T460s:~$

**Comment by maxdml** [comment]
You need to install python-bcc or python3-bcc

**Comment by disconnect3d** [comment]
> You need to install python-bcc or python3-bcc

For future readers: on Ubuntu this should be `python-bpfcc` or `python3-bpfcc`.

Also note that if you installed `bcc` module via `pip`, this has to be removed **before** installing the package from apt. If you installed from pip and then from apt, uninstall the pip package and then the apt package and install only the apt one.

**Comment by grantseltzer** [comment]
I am still getting this issue

**Comment by jackice** [comment]
I also encounter this same issue on centos8 when using the installed python3-bcc

> dnf list --installed 
> bcc.x86_64                                   0.11.0-2.el8                                      @AppStream     
> bcc-tools.x86_64                             0.11.0-2.el8                                      @AppStream     
> python3-bcc.x86_64                           0.11.0-2.el8                                      @AppStream

**Comment by C-h4ck-0** [comment]
I also encountered this error message.
You should work according to the steps listed here -
https://github.com/iovisor/bcc/blob/master/INSTALL.md

I worked by these steps (of Ubuntu in my case) and it solved the problem 👍

**Comment by C-h4ck-0** [comment]
Try running it with 'sudo su' -
![image](https://user-images.githubusercontent.com/48152831/104837149-5f8ae980-58bb-11eb-917a-e4262ba0c37f.png)

**Comment by wzhliang** [comment]
I had this problem but managed to make it disappear with `pip3 uninstall bcc`

**Comment by yueyongyue** [comment]
python2.7 ok python3 unsuccessful

**Comment by yonghong-song** [comment]
Look at Install.md file. You may need to build bcc with PYTHON_CMD option like
```
cmake .. -DPYTHON_CMD=python3 # for python3 support

```

**Comment by MihaiMaties** [comment]
FWIW - in my case it turns out that  ` sudo /bin/python3 /usr/share/bcc/tools/execsnoop` works.

python3 resolves (as per $PATH) to ~/.pyenv/shims/python3 in my case - it is 3.8.8, needed for https://github.com/elastic/rally

/bin/python3 is the "system's" i.e. CentOS 8.3 distribution's  3.6.8

**Comment by gpstrive** [comment]
use 'apt install python3-bpfcc ' ,  'pip3 install bcc' not works

**Comment by LiujiaHuan** [comment]
use 'apt install python3-bpfcc ' ,
'pip3 install bcc' not works

**Comment by minhtrex** [comment]
I tried on AlmaLinux9.
worked:  'sudo yum install python3-bcc '
not worked: 'pip3 install bcc'

**Comment by gophernancy** [comment]
I also had this problem using python3.9.
When I exec "yum install python3", it will install python3.6，it won't this problem.

**Comment by c0debreaker** [comment]
> I met a similar issue:
> 
> ```
> Traceback (most recent call last):
>   File "/path/of/bitehist.py", line 17, in <module>
>     from bcc import BPF
> ImportError: cannot import name 'BPF' from 'bcc' (/opt/python/3.9.6/lib/python3.9/site-packages/bcc/__init__.py)
> ```
> 
> And I fixed it with:
> 
> ```
> export PYTHONPATH=$(dirname `find /usr/lib -name bcc`):$PYTHONPATH
> ```
> 
> The fix is from: https://github.com/iovisor/bcc/blob/master/FAQ.txt#L17

Thanks a lot @pangle-zhang ! That's the only one that made it worked! Thank you so much!

**Comment by djh12345678** [comment]
> use 'apt install python3-bpfcc ' , 'pip3 install bcc' not works

love you bro，i try to fix this issue for three days，use this fix it！

**Comment by HEtinto** [comment]
I got this problem when use the anconda virtual python env,  i try to use command `conda deactivate ` to exit the virtual env to use the default python env to fix it.

Other problem https://github.com/bpftrace/bpftrace/issues/1855

```text
Abort on startup: CommandLine Error: Option ... registered more than once!
```

I fixed use the -DENABLE_LLVM_SHARED=1 when i build the bcc project.
(PS: i build bcc by myself.)

bcc

```ruby
root@fedora:/home/yujianming/source/bpf-learner# dnf list installed |grep bcc
bcc.x86_64                                           0.29.1-2.fc40                       @fedora          
bcc-tools.x86_64                                     0.29.1-2.fc40                       @fedora          
python3-bcc.noarch                                   0.29.1-2.fc40                       @FedeParola 
```

llvm

```ruby
root@fedora:/home/yujianming/source/bpf-learner# dnf list installed |grep llvm
llvm.x86_64                                          18.1.8-4.fc40                       @updates         
llvm-devel.x86_64                                    18.1.8-4.fc40                       @updates         
llvm-googletest.x86_64                               18.1.8-4.fc40                       @updates         
llvm-libs.x86_64                                     18.1.8-4.fc40                       @updates         
llvm-static.x86_64                                   18.1.8-4.fc40                       @updates         
llvm-test.x86_64                                     18.1.8-4.fc40                       @updates         
llvm17-libs.x86_64                                   17.0.6-7.fc40                       @fedora
```

clang

```ruby
root@fedora:/home/yujianming/source/bpf-learner# dnf list installed |grep clang
clang.x86_64                                         18.1.8-2.fc40                       @updates         
clang-devel.x86_64                                   18.1.8-2.fc40                       @updates         
clang-libs.x86_64                                    18.1.8-2.fc40                       @updates         
clang-resource-filesystem.noarch                     18.1.8-2.fc40                       @updates         
clang-tools-extra.x86_64                             18.1.8-2.fc40                       @updates         
clang17-libs.x86_64                                  17.0.6-9.fc40                       @updates         
clang17-resource-filesystem.x86_64                   17.0.6-9.fc40                       @updates
```

**Comment by 287400117** [comment]
I encountered the same issue. The problem arises because when installing python3-bcc through the package manager, it defaults to binding to the system-installed Python version directory. However, I am using a custom-installed Python version, which is why I get an error when executing python bcc. 

<img width="983" height="865" alt="Image" src="https://github.com/user-attachments/assets/b71ac6b9-b76a-44d3-989f-a0d9cfb31595" />



I fixed it with:
```
rm /usr/bin/python3
ln -s /root/.pyenv/versions/3.9.17/bin/python3 /usr/bin/python3

yum remove python3-bcc
yum install -y python3-bcc

pip3 install bcc
```

**Comment by yangzi1104** [comment]
After installing and compiling the bcc project ([using the source installation method](https://github.com/iovisor/bcc/blob/master/INSTALL.md#ubuntu---source)), you need to manually build the python package for bcc if you use the virtual environments

```
cd {your/path/to}/bcc/build/src/python/bcc-python3
python3 setup.py bdist_wheel
python3 -m pip install dist/bcc-0.35.0+b81b93e4-py3-none-any.whl
python3 -m pip list | grep bcc
```

**Comment by yangzi1104** [comment]
> After installing and compiling the bcc project ([using the source installation method](https://github.com/iovisor/bcc/blob/master/INSTALL.md#ubuntu---source)), you need to manually build the python package for bcc if you use the virtual environments
> 
> ```
> cd {your/path/to}/bcc/build/src/python/bcc-python3
> python3 setup.py bdist_wheel
> python3 -m pip install dist/bcc-0.35.0+b81b93e4-py3-none-any.whl
> python3 -m pip list | grep bcc
> ```

Another reason is that you have multiple python `virtual environments` locally. Even if you install and compile bcc in the specified virtual environment, bcc will eventually be installed in the python path of the system, usually /usr/bin/python or `/usr/bin/python3`, etc. It depends on the `-DPYTHON_CMD=python3` you specified when installing and compiling bcc, Or you can try specifying the python path of your virtual environment to `-DPYTHON_CMD` when compiling and installing bcc. You can check it through `whereis python3`.

So on the one hand, you can run it correctly through `/usr/bin/python3 hello_world.py`, or add or modify it in the file header of hello_world.py to: `#! /usr/bin/python3`, so that it can be run directly through `bash`.

Also, to verify this, you can check whether bcc has been successfully installed through `/usr/bin/python3 -m pip list | grep bcc`

**Comment by Hannah-xxl** [comment]
My OS is CentOS. I encountered this issue when install the bcc and python-bpfcc provided by the yum repository, as well as when install the bcc via pip. My final solution was to uninstall all the bcc and python-bpfcc related components from yum and pip, including all the bcc directories under the /usr path. Then, I compiled the bcc from the source code, which resolved this error finally. 
For your reference.
Here is the final successful command:
cmake.. -DCMAKE_INSTALL_PREFIX=/usr -DENABLE_LLVM_SHARED=1

I have tried to compile from source but failed yet with all the bcc components from yum or pip not removed. So maybe they cannot work around with each other, I think.
